src/index.ts

```diff
@@ -2310,8 +2304,56 @@ export class GossipSub extends EventEmitter<GossipsubEvents> implements Initiali
     this.gossipTracer.prune()
     this.publishedMessageIds.prune()
 
+    /**
+     * Instead of calling getRandomGossipPeers multiple times to:
+     *   + get more mesh peers
+     *   + more outbound peers
+     *   + oppportunistic grafting
+     *   + emitGossip
+     *
+     * We want to loop through the topic peers only single time and perform different sets at different thresholds
+     * and do emit gossip with known peers to improve performance
+     */
+
+    const peersToGossipByTopic = new Map<string, Set<PeerIdStr>>()
     // maintain the mesh for topics we have joined
     this.mesh.forEach((peers, topic) => {
+      const peersInTopic = this.topics.get(topic)
+      let candidateMeshPeers: PeerIdStr[] = []
+      const peersToGossip = new Set<PeerIdStr>()
+      peersToGossipByTopic.set(topic, peersToGossip)
+
+      if (peersInTopic) {
+        const shuffledPeers = shuffle(Array.from(peersInTopic))
+        const backoff = this.backoff.get(topic)
+        for (const id of shuffledPeers) {
+          const peerStreams = this.peers.get(id)
+          if (peerStreams && hasGossipProtocol(peerStreams.protocol) && !peers.has(id) && !this.direct.has(id)) {
+            const score = getScore(id)
+            if ((!backoff || !backoff.has(id)) && score >= 0) candidateMeshPeers.push(id)
+            if (score >= this.opts.scoreThresholds.gossipThreshold) peersToGossip.add(id)
+          }
+        }
+      }
+
+      const meshPeersFromCandidates = (ineed: number, cond: (peer: PeerIdStr) => boolean): PeerIdStr[] => {
+        let count = 0
+        let index = 0
+        const newMeshPeers = []
+
+        while (count < ineed && index < candidateMeshPeers.length) {
+          if (cond(candidateMeshPeers[index])) {
+            newMeshPeers.push(candidateMeshPeers[index])
+            candidateMeshPeers.splice(index, 1)
```

Comment:
I think splicing arrays is somewhat slow as it has too reindex all following values.

Maybe try making candidateMeshPeers a Set. It should iterate thru in insertion order (and won't need to reindex as you remove items from it)

newMeshPeers could also be a Set

Worth benchmarking and seeing

Reply:
> newMeshPeers could also be a Set

👍 

> Maybe try making candidateMeshPeers a Set

for each heartbeat we want to shuffle the peers (the random context in `getRandomGossipPeers`) so we need an ordered collection. It sounds like the `LinkedList` implementation in lodestar would be suitable here as the `splice()` method is cheap and no need to reindex